### ISSUE TITLE
Type checking is incorrect in array filtering

### ISSUE THREAD

**Issue by magiskboy** [microsoft/TypeScript#58730]
### 🔎 Search Terms

- check type in null filtering in array
- null-value is still detected by typescript although they was filtered in array

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This changed in commit or PR _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### 💻 Code

```ts
const filteredClaimIds = claimIds.filter((i) => typeof i === "string");
if (filteredClaimIds.length > 0) {
  await this.claimService.updateClaimBySubscriptionId(
    filteredClaimIds,
    {
      status: "cancelled",
    }
  );
}
```

### 🙁 Actual behavior

`this.claimService.updateClaimBySubscriptionId` requires string[] as first parameter and typescript check `filteredClaimIds` that still contains null value but I filtered in previous line. 

<img width="812" alt="image" src="https://github.com/microsoft/TypeScript/assets/13352088/82d3fff3-2045-47e0-90bd-1296cdade61b">

### 🙂 Expected behavior

Should be passed without error

### Additional information about the issue

Node version: 20.10.0
Typescript version: 5.4.5
Yarn version: 1.22.21

**Comment by MartinJohns** [comment]
You forgot to fully fill out the issue template.

Duplicate of #38390.

**Comment by jcalz** [comment]
Among other things, the template says
> Please try the nightly version of TS to see if it's already been fixed. Install typescript@next or use the [Playground](http://www.typescriptlang.org/play/?ts=Nightly).

Did you do that? I can't check because your code isn't a reproducible example, but it's quite likely to be already fixed due to [TS5.5's upcoming support for inferred type predicates](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5-beta/#inferred-type-predicates). 

Please edit and fill out the issue template properly, and if this is indeed a duplicate of #38390 and fixed in nightly, then do the TS team the favor of closing the issue so they don't have to. Thanks!
